protocols/autonatv2/src/lib.rs

```diff
@@ -0,0 +1,19 @@
+use libp2p_core::upgrade::ReadyUpgrade;
+use libp2p_swarm::StreamProtocol;
+
+pub mod client;
+mod generated;
+mod global_only;
+pub(crate) mod request_response;
+pub mod server;
+
+pub(crate) const REQUEST_PROTOCOL_NAME: StreamProtocol =
+    StreamProtocol::new("/libp2p/autonat/2/dial-request");
+pub(crate) const DIAL_BACK_PROTOCOL_NAME: StreamProtocol =
+    StreamProtocol::new("/libp2p/autonat/2/dial-back");
+pub(crate) const REQUEST_UPGRADE: ReadyUpgrade<StreamProtocol> =
+    ReadyUpgrade::new(REQUEST_PROTOCOL_NAME);
+pub(crate) const DIAL_BACK_UPGRADE: ReadyUpgrade<StreamProtocol> =
+    ReadyUpgrade::new(DIAL_BACK_PROTOCOL_NAME);
```

Comment:
I don't quite understand? What do you mean?

Reply:
Sorry for the brevity, I'll expand a bit :)

In a nutshell, my approach to software design is "write the least amount of code using the simplest primitives that get the job done". This for example translates to:

- Prefer less powerful features (functions) over more powerful features (i.e. macros) whenever you can. If you follow this principle, you end up reviewing code through the lens of "There must be a reason why this is a macro, i.e. something special is happening here and I need to pay more attention". This is also known as the [Rule of least power](https://en.wikipedia.org/wiki/Rule_of_least_power).
- Only extract code into a function / module / constant if the benefit of centralising it outweighs the cost of:
	- The indirection: When reading code, having to jump do the definition site to learn what is going on can interrupt flow.
	- The dependency: Extract code can be useful to reduce duplication but it is not free. Dependencies are a form of coupling and will make it harder to evolve code in the future because now, the requirements of all usage sites need to be considered when changing the dependency.
	- The additional code: Code you are depending on needs to be referred to somehow. `use` statements or other dependency declarations are also code that you have to read and take into account in your mental model.

In this particular case, the thing I'd argue that a constant for the protocol name itself is definitely worth it. We don't want to repeat this on in multiple places. However, the instance of `ReadyUpgrade` that uses this protocol name is not worth extracting into a constant:

- When taking into the account the definition site and the `use` statement, it is overall more lines of code to achieve the same thing.
- There is no complexity in calling a constructor. The type system ensures that you will construct a `ReadyUpgrade` and there is only one way of doing that: calling `new`.
- The chance of an error by using the wrong constant, i.e. `DIAL_BACK_UPGRADE` vs `REQUEST_UPGRADE` and `REQUEST_PROTOCOL_NAME` vs `DIAL_BACK_PROTOCOL_NAME` are the same. In both cases, we need to verify through review that we are constructing the right protocol.

Considering all this, I'd say that these constants "don't pull their weight" in that they don't have a reason to exist. In other words, they fail to bring forward sufficient arguments why they are a good design.

Hope this helps :)

Reply:
Thanks for the great explanation